lib/charms/tls_certificates_interface/v1/tls_certificates.py

```diff
@@ -0,0 +1,594 @@
+            common_name (str): Common name.
+            sans (list): Subject Alternative Name
+
+        Returns:
+            bytes: The CSR passed in the relation data
+        """
+        logger.info("Received request to create certificate")
+        relation = self.model.get_relation(self.relationship_name)
+        if not relation:
+            message = (
+                f"Relation {self.relationship_name} does not exist - "
+                f"The certificate request can't be completed"
+            )
+            logger.error(message)
+            raise RuntimeError(message)
+        relation_data = _load_relation_data(relation.data[self.model.unit])
+        csr = generate_csr(
+            private_key=private_key,
+            private_key_password=private_key_password,
+            subject=common_name,
+            sans=sans,
+        )
+        new_certificate_request = {"certificate_signing_request": csr.decode()}
+        if "certificate_signing_requests" in relation_data:
+            certificate_request_list = relation_data["certificate_signing_requests"]
+            if new_certificate_request in certificate_request_list:
+                logger.info("Request was already made - Doing nothing")
+                return csr
+            certificate_request_list.append(new_certificate_request)
+        else:
+            certificate_request_list = [new_certificate_request]
+        relation.data[self.model.unit]["certificate_signing_requests"] = json.dumps(
+            certificate_request_list
+        )
+        logger.info("Certificate request sent to provider")
+        return csr
+
+    @staticmethod
+    def _relation_data_is_valid(certificates_data: dict) -> bool:
+        """Checks whether relation data is valid based on json schema.
+
+        Args:
+            certificates_data: Certificate data in dict format.
+
+        Returns:
+            bool: Whether relation data is valid.
+        """
+        try:
+            validate(instance=certificates_data, schema=PROVIDER_JSON_SCHEMA)
+            return True
+        except exceptions.ValidationError:
+            return False
+
+    def _on_relation_changed(self, event) -> None:
+        """Handler triggerred on relation changed events.
+
+        Args:
+            event: Juju event
+
+        Returns:
+            None
+        """
+        relation_data = _load_relation_data(event.relation.data[event.unit])
+        if not self._relation_data_is_valid(relation_data):
+            logger.warning("Relation data did not pass JSON Schema validation - Deferring")
+            event.defer()
+            return
+
+        for certificate in relation_data["certificates"]:
+            self.on.certificate_available.emit(
+                certificate_signing_request=certificate["certificate_signing_request"],
+                certificate=certificate["certificate"],
+                ca=certificate["ca"],
+                chain=certificate["chain"],
```

Comment:
Maybe an expiration date can be passed here, or some simple helper function can be provided.

Reply:
Expiration date is not shown, but there are now events triggered whenever the certificate is about to expire.

Reply:
thanks, solved.